src/router.js

```diff
@@ -0,0 +1,156 @@
+import { Readable } from "stream";
+import rawBody from "raw-body";
+import { fetch, Headers, Response } from "undici";
+import { startUnleash } from "unleash-client";
+import bodyParser from "body-parser";
+
+const unleash = await startUnleash({
+  url: process.env.UNLEASH_URL ?? "",
+  appName: "router-proxy",
+  instanceId: "my-unique-instance-id",
+  customHeaders: {
+    Authorization: process.env.UNLEASH_API_TOKEN ?? "",
+  },
+});
+
+const ROLLOUT = "router-rollout";
+const REPLAY = "router-replay-on-error";
+const IGNORED_HEADERS = new Set(["connection", "content-encoding"]);
+
+/**
+ * @param {import("express").Request} req
+ */
+function useRouter(req) {
+  return unleash.isEnabled(ROLLOUT) || req.header("x-use-router") === "true";
+}
+
+function shouldReplay() {
+  return unleash.isEnabled(REPLAY);
+}
+
+/**
+ * Converts an express request into a fetch request. This consumes the request
+ * body, so we'll also return the body if we need to use it later
+ * @param {import("express").Request} req
+ * @return {Promise<[Response, Buffer]>}
+ */
+async function proxyRequest(req) {
+  const headers = new Headers();
+  for (const [key, value] of Object.entries(req.headers)) {
+    if (value && !IGNORED_HEADERS.has(key)) {
+      typeof value === "string"
+        ? headers.set(key, value)
+        : value.forEach((value) => headers.append(key, value));
+    }
+  }
+
+  // fetch takes a Buffer not a Readable stream. This consumes the request body.
+  const body = await rawBody(req);
+
+  const response = await fetch(process.env.ROUTER_URL ?? "", {
```

Comment:
Only thing here is some folks use query params as well- not sure how we'd want to represent that in this request, but nonissue.

Reply:
ah yeah that's an edge case i think we can leave as an exercise for the reader